File: src/main/scala/com/github/shadowsocks/plugin/obfs_local/ConfigFragment.scala
package com.github.shadowsocks.plugin.obfs_local

import android.os.Bundle
import android.support.v7.preference.DropDownPreference
import android.view.View
import be.mygod.preference.{EditTextPreference, PreferenceFragment}
import com.github.shadowsocks.plugin.{PluginContract, PluginOptions}

/**
  * @author Mygod
  */
class ConfigFragment extends PreferenceFragment {
  var options: PluginOptions = _

  override def onCreatePreferences(bundle: Bundle, key: String): Unit = addPreferencesFromResource(R.xml.config)

  def onInitializePluginOptions(options: PluginOptions): Unit = {
    this.options = options
    for ((key, defaultValue) <- Array(("obfs", "http"), ("obfs-host", "cloudfront.net"), ("obfs-uri", "/"))) {
      val pref = findPreference(key)
      pref match {
        case ddp: DropDownPreference => ddp.setValue(options.getOrDefault(key, defaultValue))
        case etp: EditTextPreference => etp.setText(options.getOrDefault(key, defaultValue))
      }
      pref.setOnPreferenceChangeListener((_, value) => {
        options.put(key, value.toString)
        true
      })
    }
  }

  override def onSaveInstanceState(outState: Bundle) {
    super.onSaveInstanceState(outState)
    outState.putString(PluginContract.EXTRA_OPTIONS, options.toString)
  }
  override def onViewCreated(view: View, savedInstanceState: Bundle) {
    super.onViewCreated(view, savedInstanceState)
    if (savedInstanceState != null) {
      options = new PluginOptions(savedInstanceState.getString(PluginContract.EXTRA_OPTIONS))
      onInitializePluginOptions(options)
    }
  }
}

File: app/src/main/kotlin/com/github/shadowsocks/plugin/obfs_local/ConfigFragment.kt
package com.github.shadowsocks.plugin.obfs_local

import android.os.Bundle
import androidx.preference.DropDownPreference
import androidx.preference.EditTextPreference
import androidx.preference.PreferenceFragmentCompat
import android.view.View
import com.github.shadowsocks.plugin.PluginContract
import com.github.shadowsocks.plugin.PluginOptions

/**
 * @author Mygod
 */
class ConfigFragment : PreferenceFragmentCompat() {
    var options: PluginOptions? = null

    override fun onCreatePreferences(savedInstanceState: Bundle?, rootKey: String?) {
        setPreferencesFromResource(R.xml.config, rootKey)
    }

    fun onInitializePluginOptions(pluginOptions: PluginOptions) {
        this.options = pluginOptions

        val configs = arrayOf(
            Triple("obfs", "http", null),
            Triple("obfs-host", "cloudfront.net", null),
            Triple("obfs-uri", "/", null)
        )

        for ((key, defaultValue, _) in configs) {
            val pref = findPreference<androidx.preference.Preference>(key)
            when (pref) {
                is DropDownPreference -> {
                    pref.value = pluginOptions.getOrDefault(key, defaultValue)
                }
                is EditTextPreference -> {
                    pref.text = pluginOptions.getOrDefault(key, defaultValue)
                }
            }
            pref?.setOnPreferenceChangeListener { _, newValue ->
                pluginOptions[key] = newValue.toString()
                true
            }
        }
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putString(PluginContract.EXTRA_OPTIONS, options.toString())
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        savedInstanceState?.getString(PluginContract.EXTRA_OPTIONS)?.let {
            options = PluginOptions(it)
            options?.let { opts -> onInitializePluginOptions(opts) }
        }
    }
}
